src/java/org/apache/cassandra/utils/CloseableIterator.java

```diff
@@ -24,6 +24,8 @@
 public interface CloseableIterator<T> extends Iterator<T>, AutoCloseable
 {
     public void close();
+    
+    public default void checkpoint() { }
```

Comment:
This is very awkward, although it pokes the necessary hole in the contract for RFP to be able to do what it needs to do without closing the merged iterator and therefore the merge listener. Any other ideas?

Reply:
instead of adding this to the `CloseableIterator` interface, why don't you make the anonymous MergeListener in ReplicaFilteringProtection a class that you can manipulate or "checkpoint" from `queryProtectedPartitions`?

Reply:
This is done, and if preliminary testing is any indication, it's working reasonably well.